Concretise the type inside the `GenesisHash` newtype wrapper

## Background

`GenesisHash` is defined as

```
newtype GenesisHash = GenesisHash { getGenesisHash :: forall a . Hash a }
```

To construct a value of this type one must use `coerce` from `Data.Coerce`. There's no obvious justification for this complicated approach.

## Proposed Solution

A `GenesisHash` should really be a hash of the bytes of the genesis configuration data. This is what we refer to as the genesis block. We use the type `Raw` to represent `ByteString`s that shouldn't be serialised before `hash`ing, so we should simply change the definition to use a `Hash Raw`.

> ￼ Replace `forall a . Hash a` with `Hash Raw` and fix up resulting type errors

I tried that, but that results in a type mismatch here:

```
    /home/damian/github/input-output-hk/cardano-chain/src/Cardano/Chain/Block/Header.hs:203:46: error:
        • Couldn't match type ‘cardano-binary-1.5.0:Cardano.Binary.Class.Primitive.Raw’
                         with ‘AHeader ()’
          Expected type: HeaderHash
            Actual type: Hash
                           cardano-binary-1.5.0:Cardano.Binary.Class.Primitive.Raw
        • In the second argument of ‘mkHeaderExplicit’, namely ‘prevHash’
          In the expression: mkHeaderExplicit pm prevHash difficulty
          In an equation for ‘mkHeader’:
              mkHeader pm prevHeader
                = mkHeaderExplicit pm prevHash difficulty
                where
                    prevHash = either unGenesisHash hashHeader prevHeader
                    difficulty
                      = either
                          (const 0)
                          (succ . consensusDifficulty . headerConsensusData)
                          prevHeader
        |
    203 | mkHeader pm prevHeader = mkHeaderExplicit pm prevHash difficulty
        |                                              ^^^^^^^^
```
The problem is that `hashHeader :: Header -> HeaderHash`, where `type HeaderHash = Hash Header`, so we cannot make `Hash Raw` and `Hash Header` match. 

Shouldn't we change 

```haskell
newtype GenesisHash = GenesisHash { getGenesisHash :: forall a . Hash a }
```
to 

```
newtype GenesisHash = GenesisHash { getGenesisHash :: Hash Header }
```

I guess the problem might be that the genesis hash is not really the hash of a header :/

--- a/src/Cardano/Chain/Block/Block.hs
+++ b/src/Cardano/Chain/Block/Block.hs
@@ -89,14 +89,15 @@ import Cardano.Chain.Block.Header
   , ToSign
   , decodeAHeader
   , dropBoundaryHeader
+  , genesisHeaderHash
   , hashHeader
   , headerAttributes
-  , headerProtocolVersion
   , headerDifficulty
   , headerEBDataProof
   , headerLeaderKey
   , headerPrevHash
   , headerProof
+  , headerProtocolVersion
   , headerSignature
   , headerSlot
   , headerSoftwareVersion
@@ -245,7 +246,7 @@ mkBlock
   -> Block
 mkBlock pm bv sv prevHeader = mkBlockExplicit pm bv sv prevHash difficulty
  where
-  prevHash   = either unGenesisHash hashHeader prevHeader
+  prevHash   = either genesisHeaderHash hashHeader prevHeader
   difficulty = either (const 0) (succ . headerDifficulty) prevHeader
 
 -- | Smart constructor for 'Block', without requiring the entire previous

--- a/src/Cardano/Chain/Block/Header.hs
+++ b/src/Cardano/Chain/Block/Header.hs
@@ -47,6 +47,9 @@ module Cardano.Chain.Block.Header
   , ConsensusData
   , consensusData
   , verifyConsensusData
+
+  -- * Utility functions
+  , genesisHeaderHash
   )
 where
 
@@ -54,6 +57,7 @@ import Cardano.Prelude
 
 import Control.Monad.Except (MonadError, liftEither)
 import qualified Data.ByteString as BS
+import Data.Coerce (coerce)
 import Formatting (Format, bprint, build, int)
 import qualified Formatting.Buildable as B
 
@@ -202,11 +206,15 @@ mkHeader
   -> Header
 mkHeader pm prevHeader = mkHeaderExplicit pm prevHash difficulty
  where
-  prevHash   = either unGenesisHash hashHeader prevHeader
+  prevHash   = either genesisHeaderHash hashHeader prevHeader
   difficulty = either
     (const 0)
     (succ . consensusDifficulty . headerConsensusData)
     prevHeader
+
+-- | Extract the genesis hash and cast it into a header hash.
+genesisHeaderHash :: GenesisHash -> HeaderHash
+genesisHeaderHash = coerce . unGenesisHash
 
 -- | Make a 'Header' for a given slot, with a given body, parent hash,
 --   and difficulty. This takes care of some signing and consensus data.

--- a/src/Cardano/Chain/Block/Validation.hs
+++ b/src/Cardano/Chain/Block/Validation.hs
@@ -39,7 +39,6 @@ import Cardano.Chain.Block.Block
 import Cardano.Chain.Block.Header
   ( BlockSignature(..)
   , HeaderHash
-  , genesisHeaderHash
   , recoverSignedBytes
   , wrapBoundaryBytes
   )
@@ -53,6 +52,7 @@ import Cardano.Chain.Genesis as Genesis
   , GenesisWStakeholders(..)
   , configBootStakeholders
   , configEpochSlots
+  , configGenesisHeaderHash
   , configK
   , configProtocolMagicId
   , configSlotSecurityParam
@@ -198,7 +198,7 @@ updateChainBoundary
 updateChainBoundary config cvs bvd = do
   let
     prevHash =
-      fromMaybe (genesisHeaderHash $ configGenesisHash config) (cvsPreviousHash cvs)
+      fromMaybe (configGenesisHeaderHash config) (cvsPreviousHash cvs)
 
   -- Validate the previous block hash of 'b'
   (boundaryPrevHash bvd == prevHash)
@@ -232,7 +232,7 @@ updateChain
 updateChain config cvs b = do
   let
     prevHash =
-      fromMaybe (genesisHeaderHash $ configGenesisHash config) (cvsPreviousHash cvs)
+      fromMaybe (configGenesisHeaderHash config) (cvsPreviousHash cvs)
 
   -- Validate the previous block hash of 'b'
   (blockPrevHash b == prevHash)

--- a/src/Cardano/Chain/Block/Validation.hs
+++ b/src/Cardano/Chain/Block/Validation.hs
@@ -37,7 +37,12 @@ import Cardano.Chain.Block.Block
   , blockSlot
   )
 import Cardano.Chain.Block.Header
-  (BlockSignature(..), HeaderHash, recoverSignedBytes, wrapBoundaryBytes)
+  ( BlockSignature(..)
+  , HeaderHash
+  , genesisHeaderHash
+  , recoverSignedBytes
+  , wrapBoundaryBytes
+  )
 import Cardano.Chain.Common (BlockCount(..), StakeholderId, mkStakeholderId)
 import qualified Cardano.Chain.Delegation as Delegation
 import Cardano.Chain.Delegation.Payload (APayload(..))
@@ -45,7 +50,6 @@ import Cardano.Chain.Delegation.Validation
   (delegates, initialInterfaceState, updateDelegation)
 import Cardano.Chain.Genesis as Genesis
   ( Config(..)
-  , GenesisHash(..)
   , GenesisWStakeholders(..)
   , configBootStakeholders
   , configEpochSlots
@@ -194,7 +198,7 @@ updateChainBoundary
 updateChainBoundary config cvs bvd = do
   let
     prevHash =
-      fromMaybe (unGenesisHash $ configGenesisHash config) (cvsPreviousHash cvs)
+      fromMaybe (genesisHeaderHash $ configGenesisHash config) (cvsPreviousHash cvs)
 
   -- Validate the previous block hash of 'b'
   (boundaryPrevHash bvd == prevHash)
@@ -228,7 +232,7 @@ updateChain
 updateChain config cvs b = do
   let
     prevHash =
-      fromMaybe (unGenesisHash $ configGenesisHash config) (cvsPreviousHash cvs)
+      fromMaybe (genesisHeaderHash $ configGenesisHash config) (cvsPreviousHash cvs)
 
   -- Validate the previous block hash of 'b'
   (blockPrevHash b == prevHash)

--- a/src/Cardano/Chain/Genesis/Config.hs
+++ b/src/Cardano/Chain/Genesis/Config.hs
@@ -7,6 +7,7 @@
 module Cardano.Chain.Genesis.Config
   ( StaticConfig(..)
   , Config(..)
+  , configGenesisHeaderHash
   , configK
   , configSlotSecurityParam
   , configChainQualityThreshold
@@ -37,6 +38,7 @@ import System.FilePath ((</>))
 import System.IO.Error (userError)
 
 import Cardano.Binary.Class (Raw)
+import Cardano.Chain.Block.Header (HeaderHash, genesisHeaderHash)
 import Cardano.Chain.Genesis.Data
   (GenesisData(..), GenesisDataError, readGenesisData)
 import Cardano.Chain.Genesis.Hash (GenesisHash(..))
@@ -134,6 +136,9 @@ data Config = Config
     --
     --   TODO: Figure out how to split testing and mainnet needs
     }
+
+configGenesisHeaderHash :: Config -> HeaderHash
+configGenesisHeaderHash = genesisHeaderHash . configGenesisHash
 
 configK :: Config -> BlockCount
 configK = gdK . configGenesisData

--- a/src/Cardano/Chain/Genesis/Hash.hs
+++ b/src/Cardano/Chain/Genesis/Hash.hs
@@ -8,11 +8,12 @@ where
 
 import Cardano.Prelude
 
+import Cardano.Binary.Class (Raw)
 import Cardano.Crypto.Hashing (Hash)
 
 
 newtype GenesisHash = GenesisHash
-  { unGenesisHash :: forall a. Hash a
+  { unGenesisHash :: Hash Raw
   }
 
 deriving instance Show GenesisHash

--- a/test/Test/Cardano/Chain/Elaboration/Block.hs
+++ b/test/Test/Cardano/Chain/Elaboration/Block.hs
@@ -95,7 +95,7 @@ elaborate config (_, _, pps) ast st ab = Concrete.ABlock
     }
 
   prevHash = fromMaybe
-    (Concrete.genesisHeaderHash $ Genesis.configGenesisHash config)
+    (Genesis.configGenesisHeaderHash config)
     (Concrete.cvsPreviousHash st)
 
   sid =

--- a/test/Test/Cardano/Chain/Elaboration/Block.hs
+++ b/test/Test/Cardano/Chain/Elaboration/Block.hs
@@ -95,7 +95,7 @@ elaborate config (_, _, pps) ast st ab = Concrete.ABlock
     }
 
   prevHash = fromMaybe
-    (Genesis.unGenesisHash $ Genesis.configGenesisHash config)
+    (Concrete.genesisHeaderHash $ Genesis.configGenesisHash config)
     (Concrete.cvsPreviousHash st)
 
   sid =

